### PR TITLE
Add tech group tests

### PR DESCRIPTION
**Pull request by Cjkjvfnby**
Changing tech lists in https://github.com/freeorion/freeorion/pull/2965 introduced warnings: https://github.com/freeorion/freeorion/issues/3111

@agrrr3 , @Oberlus I have changed TechLists.enqueue to eliminate warnings.  These changed do not modify what was added to the research queue. 

PS this PR is based on https://github.com/freeorion/freeorion/pull/3116

### CODE REVIEW

**Comment by geoffthemedio** [comment]
@agrrr3 @Oberlus @Morlic-fo any comments?

**Comment by agrrr3** [comment]
The fix seems dubious. In #2965 introduced one learning tech (Dep.LRN_ARTIF_MINDS_2) and one production tech (Dep.PRO_AUTO_2), both were added to TechGroup3 base in the economy dictionary. If counts are off because of #2965 it should be at those places.
The "hull" and "defense" dicts should not be involved IMHO.

**Comment by Cjkjvfnby** [comment]
> The fix seems dubious

It is not a fix.  It is how the game actually works. 

@agrrr3 , my bad, this was introduced far long before that changes.  

Nobody just noticed it for a couple of years.
https://github.com/freeorion/freeorion/pull/2338

**Comment by Cjkjvfnby** [comment]
Close and open to trigger build again.

**Comment by Cjkjvfnby** [comment]
Close and open to trigger build again.

**Comment by Cjkjvfnby** [comment]
@agrrr3 , I cannot assign you as a reviewer via UI. Can you review and leave your comments?

File: default/python/AI/TechsListsAI.py [hunk 1]
```diff
@@ -346,11 +340,8 @@ def __init__(self):
             self.economy,
             self.economy,
             Dep.LRN_ARTIF_MINDS_1,
-            self.defense,
```
Comment (agrrr3) [inline default/python/AI/TechsListsAI.py:349]: These removed self.defense entries are in TechGroup2SparseA (the others are in TechGroup2SparseB) - both inherit from TechGroup2 which does not contain a defense list, so it is ok to remove the entries.
Also in a sparse setup you do not need planetary defense soon, makes also sense for the AI.

File: default/python/AI/TechsListsAI.py [hunk 2]
```diff
@@ -467,7 +455,7 @@ def __init__(self):
             self.hull,
             self.misc,
             self.economy,
-            self.hull,
```
Comment (agrrr3) [inline default/python/AI/TechsListsAI.py:470]: TechGroup3 have asteroid reformation, heavy asteroid hulls and contragravitational maintenance techs in the self.hull list. That is not very effective. Asteroid reformation gets researched very soon. Heavy asteroid hulls are currently not a great deal compared to normal asteroid hulls. IMHO if one researches scattered asteroid hulls, heavy asteroid hull get automatically researched as well - so personally I would rather drop the heavy asteroid hulls  from TechGroup3 as well or maybe switch order for contragrav and heavy asteroid hulls and not remove this self.hull reference, but the one before (so heavy asteroid hulls are quite late to be researched).

File: default/python/AI/TechsListsAI.py [hunk 3]
```diff
@@ -522,7 +510,6 @@ def __init__(self):
...
```
Comment (agrrr3) [inline default/python/AI/TechsListsAI.py:525]: as above, rather remove the self.hull in 522

File: default/python/AI/TechsListsAI.py [hunk 4]
```diff
@@ -573,7 +560,6 @@ def __init__(self):
...
```
Comment (agrrr3) [inline default/python/AI/TechsListsAI.py:576]: as above, rather remove the entry in 573

File: default/python/AI/TechsListsAI.py (hunk 2 above)
Comment (agrrr3) [reply default/python/AI/TechsListsAI.py:470]: so, instead one could remove the entry in line 467

**Comment by agrrr3** [comment]
Ok i added some half-informed opinions. Did not have the time for checking the economy removals. Could you give a reasoning why you want to remove those self.economy entries and not other ones?

**Comment by Cjkjvfnby** [comment]
> Could you give a reasoning why you want to remove those self.economy entries and not other ones?

I just remove the latest entry, because it was not enqueued.  

Since this changed were introduced a long time ago, they can be considered a feature (I don't know author intent and nobody sends a complaint about bad AI). So I just did c cleanup, without actual changes in AI behavior.

I don't have the expertise to reason about techs, that's why I need your help.

I prefer to merge this PR as is, and do you suggested changes in the next one. 

PS.  How do you check techs? I can try to build some visualization if you need it.

**Comment by geoffthemedio** [comment]
@agrrr3 any further comments / objections...?

**Comment by agrrr3** [comment]
merge, we can do this later

**Pull request merged**